kernel-ewasm/cap9-std/src/syscalls.rs

```diff
@@ -395,20 +395,25 @@ impl Serialize for Call {
     }
 }
 
-impl Deserialize for Payload {
-    type Error = io::Error;
+impl Deserialize<u8> for Payload {
```

Comment:
I think the `Payload` and `SysCallProcedureKey` wrappers should be removed and just be inlined within each Syscall {De}Serialization.

Reply:
Any reason why? They are used quite pervasively and prevent encoding decoding errors. If we do it manually we have to worry about alignment etc. being correct at each position.

Reply:
I guess I misunderstood, looks like `Payload` doesn't include the length unlike a `Vec<u8>` being encoded.

Reply:
Thinking about it more, you're right, that was a knee-jerk reaction. Any thoughts on payload being a slice though?

Reply:
Yeah I reckon it can be a slice, particularly after fixing the comment below. Payload doesn't need to encode a length as it is just verbatim bytes, with length handled by the EVM transaction mechanism.

Was your kneejerk reaction a concern about overhead? This newtype style abstraction is guaranteed zero-overhead in Haskell, so I'd be surprised if it wasn't the same in Rust (excepting specialised functions of course). They should be mostly a feature for the typechecker that gets removed during compilation.

Reply:
By using a slice, do you mean only maintaining a reference to the original source or simply replacing `Vec<u8>` with `&[u8]`? If it's simply replacing it, I think I'm not in favour of that (in order to change to `&[u8]` we will need to be handling lifetimes at every stage right? `Payload` becomes `Payload<'a>` etc. and it's going to be heap allocated regardless).

If you're talking about maintaining a reference to the original source of the data, that might be an interesting optimisation. Although I haven't looked into the consequences fully.

I'm happy to be corrected on both these points.